common/src/main/java/dev/architectury/event/events/common/LifecycleEvent.java

```diff
@@ -0,0 +1,115 @@
+    /**
+     * Invoked before initial server startup. This is the earliest point at which the server will be available.
+     * Equivalent to Forge's {@code FMLServerAboutToStartEvent} event and
+     * Fabric's {@code ServerLifecycleEvents#SERVER_STARTING}.
+     *
+     * @see ServerState#stateChanged(Object)
+     */
+    Event<ServerState> SERVER_BEFORE_START = EventFactory.createLoop();
+    /**
+     * Invoked during server startup.
+     * Equivalent to Forge's {@code FMLServerStartingEvent} event.
+     *
+     * @see ServerState#stateChanged(Object)
+     */
+    Event<ServerState> SERVER_STARTING = EventFactory.createLoop();
+    /**
+     * Invoked when the server has started and is ready to accept players.
+     * Equivalent to Forge's {@code FMLServerStartedEvent} event
+     * and Fabric's {@code ServerLifecycleEvents#SERVER_STARTED}.
+     *
+     * @see ServerState#stateChanged(Object)
+     */
+    Event<ServerState> SERVER_STARTED = EventFactory.createLoop();
+    /**
+     * Invoked when the server begins shutting down.
+     * Equivalent to Forge's {@code FMLServerStoppingEvent} event and
+     * Fabric's {@code ServerLifecycleEvents#SERVER_STOPPING}.
+     *
+     * @see ServerState#stateChanged(Object)
+     */
+    Event<ServerState> SERVER_STOPPING = EventFactory.createLoop();
+    /**
+     * Invoked when the server has finished stopping, and is about to fully shut down.
+     * Equivalent to Forge's {@code FMLServerStoppedEvent} event and
+     * Fabric's {@code ServerLifecycleEvents#SERVER_STOPPED}.
+     *
+     * @see ServerState#stateChanged(Object)
+     */
+    Event<ServerState> SERVER_STOPPED = EventFactory.createLoop();
+    /**
+     * Invoked when a world is loaded on the server-side.
+     * Equivalent to Forge's {@code WorldEvent.Load} event (on server)
+     * and Fabric's {@code ServerWorldEvents#LOAD}.
+     *
+     * @see ServerWorldState#act(Level)
+     */
+    Event<ServerWorldState> SERVER_WORLD_LOAD = EventFactory.createLoop();
+    /**
+     * Invoked when a world is unloaded on the server-side.
+     * Equivalent to Forge's {@code WorldEvent.Unload} event (on server)
+     * and Fabric's {@code ServerWorldEvents#UNLOAD}.
+     *
+     * @see ServerWorldState#act(Level)
+     */
+    Event<ServerWorldState> SERVER_WORLD_UNLOAD = EventFactory.createLoop();
+    /**
+     * Invoked when the world is being saved.
+     * Equivalent to Forge's {@code WorldEvent.Save} event.
+     *
+     * @see ServerWorldState#act(Level)
+     */
+    Event<ServerWorldState> SERVER_WORLD_SAVE = EventFactory.createLoop();
+    
+    interface InstanceState<T> {
+        /**
+         * Parent event type for any events that are invoked on instance state change.
+         *
+         * @param instance The changed state.
+         */
+        void stateChanged(T instance);
+    }
+    
+    interface ServerState extends InstanceState<MinecraftServer> {
+    }
+    
+    interface WorldState<T extends Level> {
```

Comment:
General question: Should we keep using `World` to refer to `Level`? (as Forge is likely switching to Mojmap anyways and I doubt this counts as "mappings taint" for Yarn users due to how ubiquitous this is)